### ISSUE TITLE
Jline completion has logic issues with terminal and prompt widths (StringIndexOutOfBoundsException)

### ISSUE THREAD

**Issue by bdw429s** [jline/jline3#604]
After updating CommandBox CLI to JLine 3.17.1, some users receive this error when hitting tab:

```
java.lang.StringIndexOutOfBoundsException: String index out of range: -29
	at java.base/java.lang.String.substring(String.java:1837)
	at org.jline.reader.impl.LineReaderImpl.candidateStartPosition(LineReaderImpl.java:5192)
	at org.jline.reader.impl.LineReaderImpl.doList(LineReaderImpl.java:5048)
	at org.jline.reader.impl.LineReaderImpl.doList(LineReaderImpl.java:4998)
	at org.jline.reader.impl.LineReaderImpl.doComplete(LineReaderImpl.java:4595)
	at org.jline.reader.impl.LineReaderImpl.doComplete(LineReaderImpl.java:4365)
	at org.jline.reader.impl.LineReaderImpl.expandOrComplete(LineReaderImpl.java:4304)
	at org.jline.reader.impl.LineReaderImpl$1.apply(LineReaderImpl.java:3798)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:670)
	at org.jline.reader.impl.LineReaderImpl.readLine(LineReaderImpl.java:459)

```

After a lot of debugging, I have determined that the error happens when the **prompt is wider than the terminal width**.

There are two incorrect assumptions Jline is making.  The first is that the prompt will never be wider than the terminal width.  There is a popular module for CommandBox called Bullet Train that gives nice, pretty prompt like this, which can also be long depending on how much information is displayed:

![image](https://user-images.githubusercontent.com/584009/99582029-23b0d200-29a7-11eb-85d3-78c127d356c6.png)

The second incorrect assumption JLine is making is that there will not be any line breaks in the prompt itself.  you can see in the above image that the ACTUAL prompt is just the `> ` character on the second line.  In fact, depending on your terminal size, there can be more than one line break in the prompt string.

![image](https://user-images.githubusercontent.com/584009/99582164-5a86e800-29a7-11eb-91b1-a4e82a259e1b.png)

Jline needs to calculate the prompt length after the last line break and put in a check to ensure it is never a negative number.

**Comment by bdw429s** [comment]
@gnodet Let me know if there's anything I can do to help get the fix out for this.  I'd love to be able to squash this regression in my next release.

**Comment by mattirn** [comment]
Actually this has been already fixed in commit https://github.com/jline/jline3/commit/a5686ab138225ee8214c68bda8c33fbaf2af5d26 in case you have not enabled  JLine option AUTO_MENU_LIST (#582).

The problem is present only when option AUTO_MENU_LIST has been enabled.

**Comment by bdw429s** [comment]
@mattirn Thank you for the update.  Any idea when the fix will be available in a stable release?

**Comment by mattirn** [comment]
@bdw429s When you were planning to do your next release? Maybe we can synchronize our releases... I have not any plan for new features.

**Comment by bdw429s** [comment]
@mattirn Depending on how many users this is affecting, I may put out a patch release as soon as this fix is available.

**Comment by bdw429s** [comment]
@mattirn Hi, is there an ETA on the 3.17.2 milestone so I can plan my release?  Or perhaps a snapshot build I can use in the meantime?

**Comment by mattirn** [comment]
@bdw429s I'm on 'maintenance mode' for me 3.17.2 can be released also next week. I still would like to look if I can found a better solution for #609.  
Snapshot build you can download from [jline/3.17.2-SNAPSHOT](https://oss.sonatype.org/content/repositories/snapshots/org/jline/jline/3.17.2-SNAPSHOT/). Would be great if you can test with it before 3.17.2 will be released.

**Comment by bdw429s** [comment]
Thanks @mattirn I've pointed my build at the sonatype repo to grab 3.17.2-SNAPSHOT and I'll give it a test.

**Comment by bdw429s** [comment]
@mattirn After a quick test on `3.17.2-SNAPSHOT`, the original error does seem to be gone and it looks good.

I also activated the `AUTO_MENU_LIST` option as a test and it looks very cool.  I was able to get a divide by zero error once as it was completing but I didn't have full stack traces enabled.  After I enabled more debugging I couldn't get the divide by zero error again to save my life.  I'll keep my eye out for that one.
